components/script/dom/webgltexture.rs

```diff
@@ -416,6 +427,124 @@ impl WebGLTexture {
+            constants::R32F |
+            constants::R8UI |
+            constants::R8I |
+            constants::R16UI |
+            constants::R16I |
+            constants::R32UI |
+            constants::R32I |
+            constants::RG8 |
+            constants::RG8_SNORM |
+            constants::RG16F |
+            constants::RG32F |
+            constants::RG8UI |
+            constants::RG8I |
+            constants::RG16UI |
+            constants::RG16I |
+            constants::RG32UI |
+            constants::RG32I |
+            constants::RGB8 |
+            constants::SRGB8 |
+            constants::RGB565 |
+            constants::RGB8_SNORM |
+            constants::R11F_G11F_B10F |
+            constants::RGB9_E5 |
+            constants::RGB16F |
+            constants::RGB32F |
+            constants::RGB8UI |
+            constants::RGB8I |
+            constants::RGB16UI |
+            constants::RGB16I |
+            constants::RGB32UI |
+            constants::RGB32I |
+            constants::RGBA8 |
+            constants::SRGB8_ALPHA8 |
+            constants::RGBA8_SNORM |
+            constants::RGB5_A1 |
+            constants::RGBA4 |
+            constants::RGB10_A2 |
+            constants::RGBA16F |
+            constants::RGBA32F |
+            constants::RGBA8UI |
+            constants::RGBA8I |
+            constants::RGB10_A2UI |
+            constants::RGBA16UI |
+            constants::RGBA16I |
+            constants::RGBA32I |
+            constants::RGBA32UI |
+            constants::DEPTH_COMPONENT16 |
+            constants::DEPTH_COMPONENT24 |
+            constants::DEPTH_COMPONENT32F |
+            constants::DEPTH24_STENCIL8 |
+            constants::DEPTH32F_STENCIL8 => {},
+            _ => {
+                let is_known_compression = self
+                    .upcast::<WebGLObject>()
+                    .context()
+                    .extension_manager()
+                    .get_tex_compression_format(internal_format)
+                    .is_some();
+                if !is_known_compression {
+                    return Err(WebGLError::InvalidEnum);
+                }
+            },
+        }
+
+        let command = match target {
+            constants::TEXTURE_2D | constants::TEXTURE_CUBE_MAP => {
+                WebGLCommand::TexStorage2D(target, levels, internal_format, width, height)
+            },
+            constants::TEXTURE_3D | constants::TEXTURE_2D_ARRAY => {
+                WebGLCommand::TexStorage3D(target, levels, internal_format, width, height, depth)
+            },
+            _ => return Err(WebGLError::InvalidEnum),
+        };
+        self.upcast::<WebGLObject>().context().send_command(command);
+
+        self.initialize(
+            TexImageTarget::from_gl_constant(target).unwrap(),
+            width as u32,
+            height as u32,
+            1,
+            TexFormat::from_gl_constant(internal_format).unwrap(),
+            levels as u32,
```

Comment:
I stepped through the test failure in a debugger and realized that we are initializing a single level of the texture, but TexStorage2D expects all levels from 0 to levels to be initialized. See page 137 of https://www.khronos.org/registry/OpenGL/specs/es/3.0/es_spec_3.0.pdf#nameddest=section-3.8.4.

Reply:
Ah you're right! Indeed I should've read the spec a bit more closely, I'll do some restructuring now.